### PR TITLE
[pytorch][build][SM] PT-1.13.1 Training Release

### PR DESCRIPTION
**Pull request by ShiboXing**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.13/py3/Dockerfile.cpu [hunk 1]
```diff
@@ -124,46 +127,62 @@ RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases
...
+    # patch CVE
+    "pillow>=9.4" \
+    psutil \
+    ipython \
  && /opt/conda/bin/conda install -c dglteam -y dgl \
- # Upstream conda looks to have moved to 4.13 which is incompatible with mamba 0.22.1 and will fail the conda-forge installs.
- # having "conda update conda" before the "conda -c conda-forge" commands will automatically update conda to 4.13.
- # Moving conda update conda" after the "conda -c conda-forge" commands keep conda at 4.12 but will update other packages using
- # the current conda 4.12
- && /opt/conda/bin/conda update -y conda \
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.13/py3/Dockerfile.cpu:136]: this issue is no more?

File: pytorch/training/docker/1.13/py3/Dockerfile.cpu [hunk 2]
```diff
@@ -196,17 +215,6 @@ ARG PT_TORCHDATA_URL
 RUN pip uninstall -y torch torchvision torchaudio torchdata \
  && pip install --no-cache-dir -U ${PT_EC2_TRAINING_URL} ${PT_TORCHVISION_URL} ${PT_TORCHAUDIO_URL} ${PT_TORCHDATA_URL}
 
-RUN mkdir -p /etc/pki/tls/certs && cp /etc/ssl/certs/ca-certificates.crt /etc/pki/tls/certs/ca-bundle.crt
-
-RUN HOME_DIR=/root \
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.13/py3/Dockerfile.cpu:201]: lets keep this as is, we need to run compliance at the end for each image (ec2/sm) so that it generates file licenses for all the packages in that dockerfile section

File: pytorch/training/docker/1.13/py3/Dockerfile.cpu [hunk 3]
```diff
@@ -274,35 +276,25 @@ RUN pip install --no-cache-dir --upgrade pip --trusted-host pypi.org --trusted-h
     smclarify \
     "sagemaker>=2,<3" \
     sagemaker-experiments==0.* \
-    "sagemaker-pytorch-training<3"
+    sagemaker-pytorch-training
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.13/py3/Dockerfile.cpu:279]: do we have any implications on unpinning this from major version?

File: pytorch/training/docker/1.13/py3/Dockerfile.cpu [hunk 4]
```diff
@@ -274,35 +276,25 @@ RUN pip install --no-cache-dir --upgrade pip --trusted-host pypi.org --trusted-h
...
- && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
- && rm -rf ${HOME_DIR}/oss_compliance* \
- && rm -rf /tmp/tmp*
+RUN /opt/conda/bin/conda install -y -c conda-forge \
+  imageio \
+  shap \    
+  scikit-learn \
+  pandas \
+  && /opt/conda/bin/conda install -y -c numba numba \
+  && /opt/conda/bin/conda install -y -c anaconda seaborn \
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.13/py3/Dockerfile.cpu:295]: lets use conda-forge channel

File: test/requirements.txt [hunk 5]
```diff
@@ -2,8 +2,7 @@ docker==4.2.0
 fabric==2.5.0
 junit-xml==1.9
 packaging==20.1
-pytest==5.3.5
-pytest-rerunfailures==9.0
+pytest-rerunfailures
```
Comment (tejaschumbalkar) [inline test/requirements.txt:5]: why is pytest removed?

**Comment by ShiboXing** [comment]
SM test: last four commits

File: test/sagemaker_tests/pytorch/training/resources/fastai/cifar/cifar.py [hunk 6]
```diff
@@ -1,46 +0,0 @@
-# Copyright 2018-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/resources/fastai/cifar/cifar.py:1]: why are these files removed?

File: test/sagemaker_tests/pytorch/training/requirements.txt [hunk 7]
```diff
@@ -1,22 +1,19 @@
...
 mock==2.0.0
-pytest==4.4.0
-pytest-cov==2.7.1
-pytest-rerunfailures==8.0
+pytest
+pytest-cov
+pytest-rerunfailures
 pytest-xdist==1.28.0
 PyYAML==5.4
 requests==2.24
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/requirements.txt:16]: we do not need torch/torchvision?

**Review by tejaschumbalkar — DISMISSED** [review]

**Review by sheilaliuxl — APPROVED** [review]
LGTM for `Dockerfile.gpu` for `smmp` with `PyTorch 1.13, cu117`. Thank you for the changes!

File: test/sagemaker_tests/pytorch/training/requirements.txt (hunk 7 above)
Comment (ShiboXing) [reply test/sagemaker_tests/pytorch/training/requirements.txt:16]: Since this is the test driver, I think we wouldn't need to install torch/torchvision outside of DLC?
AWS PyTorch and torchvision have been installed in the DLC.

File: test/sagemaker_tests/pytorch/training/resources/fastai/cifar/cifar.py (hunk 6 above)
Comment (ShiboXing) [reply test/sagemaker_tests/pytorch/training/resources/fastai/cifar/cifar.py:1]: This is an old fastai test with deprecated code. Added the mnist distributed training test from upstream that test the same functionality.

**Review by tejaschumbalkar — APPROVED** [review]
